torchci/lib/bot/greenBot.ts

```diff
@@ -0,0 +1,212 @@
+import { Probot, ProbotOctokit } from "probot";
+import { reactOnComment, addLabels } from "./botUtils";
+
+const openOnGreen = new RegExp("^s*@pytorchbot\\s+?mergeOnGreen.*");
+export const LAND_PENDING = "land-pending";
+const LAND_ATTEMPT = "landed-attempt";
+const LAND_FAILED = "land-failed";
+
+type OctokitData = {
+  octokit: InstanceType<typeof ProbotOctokit>;
+  owner: string;
+  repo: string;
+  issue_number: number;
+  comment: string;
+};
+
+export default function greenBot(app: Probot): void {
+  app.on("issue_comment.created", async (ctx) => {
+    const commentBody = ctx.payload.comment.body;
+    if (commentBody.match(openOnGreen)) {
+      if (!ctx.payload.issue.pull_request) {
+        // Issue, not pull request.
+        await reactOnComment(ctx, "confused");
+      } else {
+        await reactOnComment(ctx, "+1");
+        await addLabels(ctx, [LAND_PENDING]);
+      }
+    }
+  });
+
+  app.on(["pull_request.labeled"], async (ctx) => {
+    if (ctx.payload.label.name === LAND_PENDING) {
+      // Check pull request's head ref and check runs to see if it has all green and merge if it's good
+      // TODO: Add pagination if checks go above 100
+      const payloadData = getPayloadData(ctx);
+      const ref = ctx.payload.pull_request.head.sha;
+      const prNum = ctx.payload.pull_request.number;
+      const checks = await ctx.octokit.checks.listForRef({
+        ...payloadData,
+        ref,
+        per_page: 100,
+      });
+
+      const [successfulChecks, failedChecks] = categorizeChecks(
+        checks.data.check_runs
+      );
+      // If all jobs are successful or neutral, then try merge, else do nothing
+      if (successfulChecks.length === checks.data.check_runs.length) {
+        await tryLandPR({
+          octokit: ctx.octokit,
+          ...payloadData,
+          issue_number: prNum,
+          comment: "All checks passed. Attempting to merge.",
+        });
+      }
+      // If any failed, alert the author
+      else if (failedChecks.length > 0) {
+        await failPR({
+          octokit: ctx.octokit,
+          ...payloadData,
+          issue_number: prNum,
+          comment: "Failed to land PR due to red signal.",
+        });
+      }
+    }
+  });
+
+  app.on("check_run.completed", async (ctx) => {
+    // Check that there are associated PRs that have the label
+    const prWithLabelsAwaitable = ctx.payload.check_run.pull_requests.map(
+      (pr) => {
+        return ctx.octokit.pulls.get({
+          ...getPayloadData(ctx),
+          pull_number: pr.number,
+        });
+      }
+    );
+    const prs = await Promise.all(prWithLabelsAwaitable);
+    const landPendingPrs = prs.filter((pr) => {
+      const ind = pr.data.labels.find((label) => {
+        return label.name === LAND_PENDING;
+      });
+      return ind != null;
+    });
+
+    if (landPendingPrs.length === 0) {
+      return;
+    }
+
+    if (landPendingPrs.length > 1) {
```

Comment:
What is actually in the `check_run.pull_requests` field? How could there be more than one thing?

Reply:
I'm not really sure. When I run this query in rockset
```
select
    pull_requests
from
    commons.workflow_run w
where
    length(pull_requests) > 1
order by
    _event_time desc
```
I get a decent amount of results and they usually are from pull requests like this https://github.com/luojiguicai/pytorch/pull/39 where they're trying to pull in changes from pytorch/pytorch into their master.

Reply:
Hmmmm…the [payload docs](https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#webhook-payload-object-1) for this suggest that this approach won't work for forked pull requests:
 
> When the check suite's head_branch is in a forked repository it will be null and the pull_requests array will be empty.

We might need another way of figuring this out.